File: comms/server/api/conn.py
################################################################################
# A network connection
################################################################################
# Copyright: IQGeo Limited 2010-2023

from copy import copy
from myworldapp.core.server.base.core.myw_error import MywError
from myworldapp.core.server.dd.myw_reference import MywReference
from .pin_range import PinRange
from .network import Network


class Conn:
    """
    A connection from one set of pins to another

    Deals with the business of reversing connection when looking
    upstream"""

    # ------------------------------------------------------------------------------
    #                                   CONSTRUCTION
    # ------------------------------------------------------------------------------

    def __init__(self, conn_rec, forward=True):
        """
        Init slots of self from connection record CONN_REC

        If FORWARD is false, reverse the connection"""

        # Init slots
        self.conn_rec = conn_rec
        self.forward = forward
        self.network = Network.connection_types[conn_rec.feature_type]

        # Get from and to info
        if forward == True:
            from_urn = conn_rec.in_object
            from_pin_side = conn_rec.in_side
            from_pin_low = conn_rec.in_low
            from_pin_high = conn_rec.in_high
            to_urn = conn_rec.out_object
            to_pin_side = conn_rec.out_side
            to_pin_low = conn_rec.out_low
            to_pin_high = conn_rec.out_high

        elif forward == False:
            from_urn = conn_rec.out_object
            from_pin_side = conn_rec.out_side
            from_pin_low = conn_rec.out_low
            from_pin_high = conn_rec.out_high
            to_urn = conn_rec.in_object
            to_pin_side = conn_rec.in_side
            to_pin_low = conn_rec.in_low
            to_pin_high = conn_rec.in_high

        else:
            raise MywError("Bad direction:", forward)

        self.from_ref = MywReference.parseUrn(from_urn)
        self.to_ref = MywReference.parseUrn(to_urn)
        self.from_pins = PinRange(from_pin_side, from_pin_low, from_pin_high)
        self.to_pins = PinRange(to_pin_side, to_pin_low, to_pin_high)

    def intersect(self, from_pins):
        """
        The subset of self that relates to FROM_PINS (if any)

        Returns a Conn (or None)"""

        # Find pin range overlap
        overlap = self.from_pins.intersect(from_pins)
        if not overlap:
            return None

        # Return a shallow copy of self with pins updated
        # ENH: Use contructor instead
        conn = copy(self)

        conn.from_pins = overlap

        conn.to_pins = PinRange(
            self.to_pins.side, self.toPinFor(overlap.low), self.toPinFor(overlap.high)
        )

        return conn

    # ------------------------------------------------------------------------------
    #                                    PROPERTIES
    # ------------------------------------------------------------------------------

    def __str__(self):
        """
        String representation of self for debug messages etc
        """

        return "{}({})".format(self.__class__.__name__, self.spec)

    @property
    def spec(self):
        """
        String representation of self's data
        """

        return "{}#{} -> {}#{}".format(
            self.from_ref.urn(), self.from_pins.spec, self.to_ref.urn(), self.to_pins.spec
        )

    @property
    def db_view(self):
        """
        FeatureView that self's record comes from
        """

        return self.conn_rec._view

    def description(self):
        """
        String representation of self for GUI
        """

        sep = "->" if self.forward else "<-"

        return "{} {} {}".format(self.fromDescription(), sep, self.toDescription())

    def fromDescription(self):
        """
        String representation of self's from side for GUI
        """

        obj = self.fromFeatureRec()

        if not obj:
            return "Bad reference: {}#{}".format(self.from_ref, self.from_pins.spec)

        if self.is_from_cable:
            obj = obj._field("cable").rec()

        return "{}#{}".format(obj.name, self.from_pins.spec)

    def toDescription(self):
        """
        String representation of self's to side for GUI
        """

        obj = self.toFeatureRec()

        if not obj:
            return "Bad reference: {}#{}".format(self.to_ref, self.to_pins.spec)

        if self.is_to_cable:
            obj = obj._field("cable").rec()

        return "{}#{}".format(obj.name, self.to_pins.spec)

    @property
    def is_from_cable(self):
        """
        True if self runs from a cable segment
        """

        return self.from_feature_type == self.network.segment_type

    @property
    def is_to_cable(self):
        """
        True if self runs to a cable segment
        """

        return self.to_feature_type == self.network.segment_type

    @property
    def is_splice(self):
        """
        True if self connects two cables
        """

        return self.is_from_cable and self.is_to_cable

    @property
    def to_feature_type(self):
        """
        The feature type that self connects to
        """

        return self.to_ref.feature_type

    @property
    def from_feature_type(self):
        """
        The feature type that self connects to
        """

        return self.from_ref.feature_type

    def toFeatureRec(self):
        """
        The feature record to which the connection points
        """

        return self.db_view.get(self.to_ref)

    def fromFeatureRec(self):
        """
        The feature record from which the connection points
        """

        return self.db_view.get(self.from_ref)

    # ------------------------------------------------------------------------------
    #                                   PIN MAPPING
    # ------------------------------------------------------------------------------

    def pinPairs(self):
        """
        Yields from,to pairs of self
        """

        # ENH: Check range sizes match

        to_pin = self.to_pins.low
        for from_pin in self.from_pins.range():
            yield from_pin, to_pin
            to_pin += 1

    def toPinsFor(self, from_pins):
        """
        The pins that FROM_PINS connect to

        Assumes FROM_PINS is within range"""

        return PinRange(
            self.to_pins.side, self.toPinFor(from_pins.low), self.toPinFor(from_pins.high)
        )

    def toPinFor(self, from_pin):
        """
        The pin that FROM_PIN connects to

        Assumes FROM_PIN is within range"""

        return self.to_pins.low + (from_pin - self.from_pins.low)

    def fromPinFor(self, to_pin):
        """
        The pin that TO_PIN connects to

        Assumes TO_PIN is within range"""

        return self.from_pins.low + (to_pin - self.to_pins.low)

    # ------------------------------------------------------------------------------
    #                                   SERIALISATION
    # ------------------------------------------------------------------------------

    def definition(self):
        """
        Self in JSON-serialisable form
        """

        defn = {}

        defn["urn"] = self.conn_rec._urn()
        defn["forward"] = self.forward
        defn["from_feature"] = self.from_ref.urn()
        defn["from_pins"] = self.from_pins.definition()
        defn["to_feature"] = self.to_ref.urn()
        defn["to_pins"] = self.to_pins.definition()

        # For cable segments, add cable
        if self.is_from_cable:
            defn["from_cable"] = self.fromFeatureRec().cable
            defn["from_cable_side"] = self.from_pins.otherSide()

        if self.is_to_cable:
            defn["to_cable"] = self.toFeatureRec().cable
            defn["to_cable_side"] = self.to_pins.otherSide()

        # For delta records, add name of owner
        # ENH: Duplicated with myw_feature_model_mixin
        # ENH: Make optional
        delta = self.db_view.delta
        if delta is not None:
            delta_owner = self.db_view.get(delta)
            defn["delta"] = {"name": delta, "title": delta_owner._title()}

        return defn

    def asRec(self, **extra_fields):
        """
        Self as a pseudo-record

        Used for export"""

        rec = {**extra_fields}
        rec["in_object"] = self.from_ref.urn()
        rec["in_side"] = self.from_pins.side
        rec["in_low"] = self.from_pins.low
        rec["in_high"] = self.from_pins.high
        rec["out_object"] = self.to_ref.urn()
        rec["out_side"] = self.to_pins.side
        rec["out_low"] = self.to_pins.low
        rec["out_high"] = self.to_pins.high
        rec["housing"] = self.conn_rec.housing

        return rec

    def features(self):
        """
        The features referenced in self

        Returns a set of features"""

        features = set()

        features.add(self.fromFeatureRec())
        features.add(self.toFeatureRec())

        # For cable segments, add cable
        if self.is_from_cable:
            features.add(self.fromFeatureRec()._field("cable").rec())

        if self.is_to_cable:
            features.add(self.toFeatureRec()._field("cable").rec())

        return features

File: comms/native/services/api/Conn.js
//###############################################################################
// A network connection
//###############################################################################
// Copyright: IQGeo Limited 2010-2023

//import OrderedDict from 'collections'
//import copy from 'copy'
//import {MywError} from 'myWorld-native-services'

import PinRange from './PinRange';
import Network from './Network';
import { MywError, Reference } from 'myWorld-native-services';
import { MywClass } from 'myWorld-base';

/**
 * A connection from one set of pins to another
 *
 * Deals with the business of reversing connection when looking
 * upstream
 */
class Conn extends MywClass {
    // ------------------------------------------------------------------------------
    //                                   CONSTRUCTION
    // ------------------------------------------------------------------------------

    /**
     * Init slots of self from connection record CONN_REC
     *
     * If FORWARD is false, reverse the connection
     */
    constructor(conn_rec, forward = true) {
        super();
        // Init slots
        this.conn_rec = conn_rec;
        this.db_view = conn_rec.view;
        this.forward = forward;
        this.network = Network.connection_types[conn_rec.getType()];

        let from_urn,
            from_pin_side,
            from_pin_low,
            from_pin_high,
            to_urn,
            to_pin_side,
            to_pin_low,
            to_pin_high;

        // Get from and to info
        if (forward == true) {
            from_urn = conn_rec.properties.in_object;
            from_pin_side = conn_rec.properties.in_side;
            from_pin_low = conn_rec.properties.in_low;
            from_pin_high = conn_rec.properties.in_high;
            to_urn = conn_rec.properties.out_object;
            to_pin_side = conn_rec.properties.out_side;
            to_pin_low = conn_rec.properties.out_low;
            to_pin_high = conn_rec.properties.out_high;
        } else if (forward == false) {
            from_urn = conn_rec.properties.out_object;
            from_pin_side = conn_rec.properties.out_side;
            from_pin_low = conn_rec.properties.out_low;
            from_pin_high = conn_rec.properties.out_high;
            to_urn = conn_rec.properties.in_object;
            to_pin_side = conn_rec.properties.in_side;
            to_pin_low = conn_rec.properties.in_low;
            to_pin_high = conn_rec.properties.in_high;
        } else {
            throw MywError('Bad direction:', forward);
        }

        this.from_ref = Reference.parseUrn(from_urn);
        this.to_ref = Reference.parseUrn(to_urn);
        this.from_pins = new PinRange(from_pin_side, from_pin_low, from_pin_high);
        this.to_pins = new PinRange(to_pin_side, to_pin_low, to_pin_high);
    }

    /**
     * The subset of self that relates to FROM_PINS (if any)
     *
     * Returns a Conn (or None)
     */
    intersect(from_pins) {
        // Find pin range overlap
        const overlap = this.from_pins.intersect(from_pins);
        if (!overlap) {
            return undefined;
        }

        // Return a shallow copy of self with pins updated
        const conn = new Conn(this.conn_rec, this.forward);

        conn.from_pins = overlap;

        conn.to_pins = new PinRange(
            this.to_pins.side,
            this.toPinFor(overlap.low),
            this.toPinFor(overlap.high)
        );

        return conn;
    }

    // ------------------------------------------------------------------------------
    //                                    PROPERTIES
    // ------------------------------------------------------------------------------

    /**
     * String representation of self for debug messages etc
     */
    __ident__() {
        return `Conn(${this.spec})`;
    }

    //@property
    /**
     * String representation of self's data
     */
    _spec() {
        return `${this.from_ref.urn()}/${this.from_pins.spec} -> ${this.to_ref.urn()}/${
            this.to_pins.spec
        }`;
    }

    /**
     * String representation of self for GUI
     */
    async description() {
        const sep = this.forward ? '->' : '<-';
        const fromDescription = await this.fromDescription();
        const toDescription = await this.toDescription();

        return `${fromDescription} ${sep} ${toDescription}`;
    }

    /**
     * String representation of this's from side for GUI
     */
    async fromDescription() {
        let obj = await this.fromFeatureRec();
        if (this.is_from_cable) obj = await obj.followRef('cable');

        return `${obj.properties.name}#${this.from_pins.spec}`;
    }

    /**
     * String representation of this's to side for GUI
     */
    async toDescription() {
        let obj = await this.toFeatureRec();
        if (this.is_to_cable) obj = await obj.followRef('cable');

        return `${obj.properties.name}#${this.to_pins.spec}`;
    }

    //@property
    /**
     * True if self runs from a cable segment
     */
    _is_from_cable() {
        return this.from_feature_type == this.network.segment_type;
    }

    //@property
    /**
     * True if self runs to a cable segment
     */
    _is_to_cable() {
        return this.to_feature_type == this.network.segment_type;
    }

    //@property
    /**
     * True if self connects two cables
     */
    _is_splice() {
        return this.is_from_cable && this.is_to_cable;
    }

    //@property
    /**
     * The feature type that self connects to
     */
    _to_feature_type() {
        return this.to_ref.feature_type;
    }

    //@property
    /**
     * The feature type that self connects to
     */
    _from_feature_type() {
        return this.from_ref.feature_type;
    }

    /**
     * The feature record to which the connection points
     */
    async toFeatureRec() {
        if (this.forward) {
            return this.conn_rec.followRef('out_object');
        } else {
            return this.conn_rec.followRef('in_object');
        }
    }

    /**
     * The feature record from which the connection points
     */
    async fromFeatureRec() {
        if (this.forward) {
            return this.conn_rec.followRef('in_object');
        } else {
            return this.conn_rec.followRef('out_object');
        }
    }

    // ------------------------------------------------------------------------------
    //                                   PIN MAPPING
    // ------------------------------------------------------------------------------

    /**
     * Yields from,to pairs of self
     */
    pinPairs() {
        let to_pin = this.to_pins.low;
        const from_pin_range = this.from_pins.range();
        const pairs = [];

        for (const from_pin of from_pin_range) {
            pairs.push([from_pin, to_pin]);
            to_pin += 1;
        }

        return pairs;
    }

    /**
     * The pins that FROM_PINS connect to
     *
     * Assumes FROM_PINS is within range
     */
    toPinsFor(from_pins) {
        return new PinRange(
            this.to_pins.side,
            this.toPinFor(from_pins.low),
            this.toPinFor(from_pins.high)
        );
    }

    /**
     * The pin that FROM_PIN connects to
     *
     * Assumes FROM_PIN is within range
     */
    toPinFor(from_pin) {
        return this.to_pins.low + (from_pin - this.from_pins.low);
    }

    /**
     * The pin that TO_PIN connects to
     *
     * Assumes TO_PIN is within range
     */
    fromPinFor(to_pin) {
        return this.from_pins.low + (to_pin - this.to_pins.low);
    }

    // ------------------------------------------------------------------------------
    //                                   SERIALISATION
    // ------------------------------------------------------------------------------

    /**
     * Self in JSON-serialisable form
     */
    async definition() {
        const defn = {};

        defn['urn'] = this.conn_rec.getUrn();
        defn['forward'] = this.forward;
        defn['from_feature'] = this.from_ref.urn();
        defn['from_pins'] = this.from_pins.definition();
        defn['to_feature'] = this.to_ref.urn();
        defn['to_pins'] = this.to_pins.definition();

        // For cable segments, add cable
        if (this.is_from_cable) {
            const from_rec = await this.fromFeatureRec();
            defn['from_cable'] = from_rec.properties.cable;
            defn['from_cable_side'] = this.from_pins.otherSide();
        }

        if (this.is_to_cable) {
            const to_rec = await this.toFeatureRec();
            defn['to_cable'] = to_rec.properties.cable;
            defn['to_cable_side'] = this.to_pins.otherSide();
        }

        // get delta owner description for displaying proposed objects
        // ENH: this is duplicated from myw_feature_model_mixin
        const delta = this.db_view.delta;
        if (delta) {
            const delta_owner = await this.db_view.get(delta);
            defn['delta'] = { name: delta, title: delta_owner.myw.title };
        }

        return defn;
    }

    /**
     * The features referenced in self
     *
     * Returns a set of features
     */
    features() {
        const features = set(); // eslint-disable-line

        features.add(this.fromFeatureRec());
        features.add(this.toFeatureRec());

        // For cable segments, add cable
        if (this.is_from_cable) {
            features.add(this.fromFeatureRec()._field('cable').rec());
        }

        if (this.is_to_cable) {
            features.add(this.toFeatureRec()._field('cable').rec());
        }

        return features;
    }
}

Object.defineProperty(Conn.prototype, 'spec', {
    get() {
        return this._spec();
    }
});

Object.defineProperty(Conn.prototype, 'is_from_cable', {
    get() {
        return this._is_from_cable();
    }
});

Object.defineProperty(Conn.prototype, 'is_to_cable', {
    get() {
        return this._is_to_cable();
    }
});

Object.defineProperty(Conn.prototype, 'is_splice', {
    get() {
        return this._is_splice();
    }
});

Object.defineProperty(Conn.prototype, 'to_feature_type', {
    get() {
        return this._to_feature_type();
    }
});

Object.defineProperty(Conn.prototype, 'from_feature_type', {
    get() {
        return this._from_feature_type();
    }
});

export default Conn;
